packages/discount-code-importer/src/main.js

```diff
@@ -1,3 +1,221 @@
+    this.client = createClient({
+      middlewares: [
+        createAuthMiddlewareForClientCredentialsFlow(this.apiConfig),
+        createUserAgentMiddleware({
+          libraryName: 'discount-code-importer',
+          libraryVersion: version,
+        }),
+        createHttpMiddleware({ host: this.apiConfig.host }),
+      ],
+    })
+
+    this.syncDiscountCodes = createSyncDiscountCodes()
+
+    this.logger = logger || {
+      error: npmlog.error.bind(this, ''),
+      warn: npmlog.warn.bind(this, ''),
+      info: npmlog.info.bind(this, ''),
+      verbose: npmlog.verbose.bind(this, ''),
+    }
+    this._resetSummary()
+  }
+
+  // Use static method because this is not called on any instance
+  static _buildPredicate (codeObjects: CodeDataArray): string {
+    const predicateArray = _.map(codeObjects, 'code')
+    return `code in ("${predicateArray.join('", "')}")`
+  }
+
+  // Wrapper function for compatibility with CLI
+  processStream (chunk: ChunkOptions, cb: () => mixed) {
+    this.logger.info('Starting conversion')
+    return this._processBatches(chunk)
+      .then(() => cb())
+      // No catch block as errors will be caught in the CLI
+  }
+
+  _processBatches (codes: CodeDataArray) {
+    // Batch to `batchSize` to reduce necessary fetch API calls
+    const batchedList = _.chunk(codes, this.batchSize)
+    return Promise.mapSeries(batchedList, (codeObjects: CodeDataArray) => {
+      // Build predicate and fetch existing code
+      const predicate = DiscountCodeImport._buildPredicate(codeObjects)
+      const service = this._createService()
+      const uri = service
+        .where(predicate)
+        .perPage(this.batchSize)
+        .build()
+      return this.client.execute({
+        uri,
+        method: 'GET',
+      })
+        .then(({ body: { results: existingCodes } }: Object) => (
+          this._createOrUpdate(codeObjects, existingCodes)
+        ))
+    })
+      .then(() => Promise.resolve(this.summaryReport()))
+      .catch(error => Promise.reject({
+        error,
+        summary: this._summary,
+      }),
+      )
+  }
+
+  _createOrUpdate (newCodes: CodeDataArray, existingCodes: CodeDataArray) {
+    return Promise.map(newCodes, (newCode: CodeData) => {
+      const existingCode = _.find(existingCodes, ['code', newCode.code])
+      if (existingCode)
+        return this._update(newCode, existingCode)
+          .then((response) => {
+            if (response && response.statusCode === 304)
+              this._summary.unchanged += 1
+            else
+              this._summary.updated += 1
+            return Promise.resolve()
+          })
+          .catch((error) => {
+            if (this.continueOnProblems) {
+              this._summary.updateErrorCount += 1
+              this._summary.errors.push(error)
+              // eslint-disable-next-line max-len
+              const msg = 'Update error occured but ignored. See summary for details'
+              this.logger.warn(msg)
```

Comment:
Let us log the error here also with `this.logger`. Very useful is logs here is being transported to another medium